### PR TITLE
[GPU] Updated to allocate memory in order of size while deserializing

### PR DESCRIPTION
**Pull request by e-ddykim**
### Details:
 - This PR changes to allocate memory in order from largest to smallest when deserializing.
 - This is the same order as when running without cache.

### Tickets:
 - 107612

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 1]
```diff
@@ -549,14 +554,35 @@ void network::save(cldnn::BinaryOutputBuffer& ob) {
...
+
+    std::vector<std::shared_ptr<primitive_inst>> insts_to_allocate(_exec_order.begin(), _exec_order.end());
+    std::sort(insts_to_allocate.begin(),
+              insts_to_allocate.end(),
+              [&exec_order_num](std::shared_ptr<primitive_inst> const& lhs, std::shared_ptr<primitive_inst> const& rhs) {
+                    long lhs_size = (lhs->mem_allocated()) ? lhs->get_output_layout().bytes_count() : exec_order_num[lhs->id()];
+                    long rhs_size = (rhs->mem_allocated()) ? rhs->get_output_layout().bytes_count() : exec_order_num[rhs->id()];
+                    return (lhs_size > rhs_size);
+              });
+
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/network.cpp:571]: Seems that it's sorting twice, once in the allocate_instance phase, and then here in save. 
Could we store the first sort result and then just use it here?

File: src/plugins/intel_gpu/src/graph/data.cpp [hunk 2]
```diff
@@ -92,7 +92,7 @@ void data_inst::load(BinaryInputBuffer& ib) {
         pos += data_size;
         ib.seekg(pos);
     } else {
-        _outputs[0] = get_network().get_memory_pool().get_memory(output_layout, _allocation_type, false);
+        _outputs[0] = get_network().get_engine().allocate_memory(output_layout, _allocation_type, false);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/data.cpp:95]: Is this change has any impact?

File: src/plugins/intel_gpu/src/graph/network.cpp (hunk 1 above)
Comment (e-ddykim) [reply src/plugins/intel_gpu/src/graph/network.cpp:571]: They are different in that this sorting logic moves primitive_insts that are not need to allocate memory to the end of the list.
This is necessary to reuse memory in these primitive_insts.

File: src/plugins/intel_gpu/src/graph/data.cpp (hunk 2 above)
Comment (e-ddykim) [reply src/plugins/intel_gpu/src/graph/data.cpp:95]: Actually, this change has no impact on performance and any functional behavior.
The purpose of this change is to allocate constant data memory with the same procedure as when not using the cache.

**Review by yeonbok — APPROVED** [review]

**Pull request merged**